File: src/algorithm/evaluation/section_marking_scheme.py
from copy import deepcopy

from src.algorithm.evaluation.answer_matcher import AnswerMatcher
from src.schemas.constants import (
    BONUS_SECTION_PREFIX,
    DEFAULT_SECTION_KEY,
    SCHEMA_VERDICTS_IN_ORDER,
    VERDICT_TO_SCHEMA_VERDICT,
    VERDICTS_IN_ORDER,
    MarkingSchemeType,
    SchemaVerdict,
    Verdict,
)
from src.utils.logger import logger
from src.utils.parsing import parse_fields


class SectionMarkingScheme:
    def __init__(self, section_key, section_scheme, set_name, empty_value):
        # TODO: get local empty_value from qblock
        self.empty_value = empty_value
        self.section_key = section_key
        self.set_name = set_name
        self.marking_type = section_scheme.get("marking_type", "default")
        self.reset_all_streaks()

        # DEFAULT marking scheme follows a shorthand
        if section_key == DEFAULT_SECTION_KEY:
            self.questions = None
            self.marking = self.parse_verdict_marking_from_scheme(section_scheme)
        else:
            self.questions = parse_fields(section_key, section_scheme["questions"])
            self.marking = self.parse_verdict_marking_from_scheme(
                section_scheme["marking"]
            )
        self.validate_marking_scheme()

    def reset_all_streaks(self):
        # TODO: support for MarkingSchemeType.SECTION_LEVEL_STREAK
        if self.marking_type == MarkingSchemeType.VERDICT_LEVEL_STREAK:
            self.streaks = {
                schema_verdict: 0 for schema_verdict in SCHEMA_VERDICTS_IN_ORDER
            }
        else:
            self.section_level_streak = 0
            self.previous_streak_verdict = None

    def get_delta_for_verdict(
        self, answer_matcher_marking, question_verdict, current_streak
    ):
        if type(answer_matcher_marking[question_verdict]) == list:
            return answer_matcher_marking[question_verdict][current_streak]
        else:
            if current_streak > 0:
                logger.warning(
                    f"Non zero streak({current_streak}) for verdict {question_verdict} in scheme {self}. Using non-streak score for this verdict."
                )
            return answer_matcher_marking[question_verdict]

    def get_delta_and_update_streak(
        self, answer_matcher_marking, answer_type, question_verdict, allow_streak
    ):
        schema_verdict = AnswerMatcher.get_schema_verdict(
            answer_type, question_verdict, 0
        )

        # TODO: support for MarkingSchemeType.SECTION_LEVEL_STREAK

        if self.marking_type == MarkingSchemeType.VERDICT_LEVEL_STREAK:
            current_streak = self.streaks[schema_verdict]
            # reset all
            self.reset_all_streaks()

            if allow_streak:
                # increase only current verdict streak
                if schema_verdict != SchemaVerdict.UNMARKED:
                    self.streaks[schema_verdict] = current_streak + 1

            delta = self.get_delta_for_verdict(
                answer_matcher_marking, question_verdict, current_streak
            )
            updated_streak = self.streaks[schema_verdict]

        elif self.marking_type == MarkingSchemeType.SECTION_LEVEL_STREAK:
            # TODO: add tests for this type

            # Note: this is assuming that the parent is calling in order of section question.
            current_streak = self.section_level_streak
            previous_verdict = self.previous_streak_verdict
            # reset all
            self.reset_all_streaks()

            if allow_streak:
                # increase only current verdict streak
                if previous_verdict is None or schema_verdict == previous_verdict:
                    self.section_level_streak = current_streak + 1

            delta = self.get_delta_for_verdict(
                answer_matcher_marking, question_verdict, current_streak
            )
            updated_streak = self.section_level_streak
        else:
            current_streak, updated_streak = 0, 0
            delta = self.get_delta_for_verdict(
                answer_matcher_marking, question_verdict, current_streak
            )

        return delta, current_streak, updated_streak

    def deepcopy_with_questions(self, questions):
        clone = deepcopy(self)
        clone.update_questions(questions)
        return clone

    def update_questions(self, questions):
        self.questions = questions
        self.validate_marking_scheme()

    def validate_marking_scheme(self):
        # TODO: add validation on maximum streak possible vs length of provided section verdict marking
        pass

    def __str__(self):
        return self.section_key

    def parse_verdict_marking_from_scheme(self, section_scheme):
        parsed_marking = {}
        for verdict in VERDICTS_IN_ORDER:
            schema_verdict = VERDICT_TO_SCHEMA_VERDICT[verdict]
            schema_verdict_marking = AnswerMatcher.parse_verdict_marking(
                section_scheme[schema_verdict]
            )
            if (
                self.marking_type == MarkingSchemeType.DEFAULT
                and schema_verdict_marking > 0
                and schema_verdict == SchemaVerdict.INCORRECT
                and not self.section_key.startswith(BONUS_SECTION_PREFIX)
            ):
                logger.warning(
                    f"Found positive marks({round(schema_verdict_marking, 2)}) for incorrect answer in the schema '{self.section_key}'. For Bonus sections, prefer adding a prefix 'BONUS_' to the scheme name."
                )

            # translated marking e.g. parsed[MATCHED] -> section_scheme['correct']
            parsed_marking[verdict] = schema_verdict_marking

        return parsed_marking

    def get_bonus_type(self):
        if self.marking_type == MarkingSchemeType.VERDICT_LEVEL_STREAK:
            return None
        elif self.marking[Verdict.NO_ANSWER_MATCH] <= 0:
            return None
        elif self.marking[Verdict.UNMARKED] > 0:
            return "BONUS_FOR_ALL"
        elif self.marking[Verdict.UNMARKED] == 0:
            return "BONUS_ON_ATTEMPT"
        else:
            return None

File: omrchecker-js/packages/core/src/processors/evaluation/SectionMarkingScheme.ts
/**
 * Section marking scheme - Handles scoring rules for question sections
 * Port of Python's src/processors/evaluation/section_marking_scheme.py
 */

import { AnswerMatcher, AnswerType, SchemaVerdict, Verdict } from './AnswerMatcher';
import { Logger } from '../../utils/logger';
import { deepClone } from '../../utils/object';

const logger = new Logger('SectionMarkingScheme');

/**
 * Marking scheme types
 */
export enum MarkingSchemeType {
  DEFAULT = 'default',
  VERDICT_LEVEL_STREAK = 'verdict_level_streak',
  SECTION_LEVEL_STREAK = 'section_level_streak',
}

/**
 * Default section key
 */
export const DEFAULT_SECTION_KEY = 'DEFAULT';

/**
 * Bonus section prefix
 */
export const BONUS_SECTION_PREFIX = 'BONUS_';

/**
 * Section marking scheme configuration
 */
export interface SectionSchemeConfig {
  marking_type?: string;
  questions?: string[];
  marking?: {
    correct: number | number[];
    incorrect: number | number[];
    unmarked: number | number[];
  };
  correct?: number | number[];
  incorrect?: number | number[];
  unmarked?: number | number[];
}

/**
 * Section marking scheme - manages scoring for a section of questions
 */
export class SectionMarkingScheme {
  public sectionKey: string;
  public setName: string;
  public emptyValue: string;
  public markingType: MarkingSchemeType;
  public questions: string[] | null;
  public marking: Record<string, number>;

  // Streak tracking
  private streaks: Record<SchemaVerdict, number>;
  private sectionLevelStreak: number;
  private previousStreakVerdict: SchemaVerdict | null;

  constructor(
    sectionKey: string,
    sectionScheme: SectionSchemeConfig,
    setName: string,
    emptyValue: string
  ) {
    this.sectionKey = sectionKey;
    this.setName = setName;
    this.emptyValue = emptyValue;
    this.markingType = (sectionScheme.marking_type as MarkingSchemeType) || MarkingSchemeType.DEFAULT;
    this.streaks = {} as Record<SchemaVerdict, number>;
    this.sectionLevelStreak = 0;
    this.previousStreakVerdict = null;

    this.resetAllStreaks();

    // DEFAULT marking scheme follows a shorthand
    if (sectionKey === DEFAULT_SECTION_KEY) {
      this.questions = null;
      this.marking = this.parseVerdictMarkingFromScheme(sectionScheme);
    } else {
      this.questions = sectionScheme.questions || [];
      const markingConfig = sectionScheme.marking || sectionScheme;
      this.marking = this.parseVerdictMarkingFromScheme(markingConfig);
    }

    this.validateMarkingScheme();
  }

  /**
   * Reset all streak counters
   */
  resetAllStreaks(): void {
    if (this.markingType === MarkingSchemeType.VERDICT_LEVEL_STREAK) {
      this.streaks = {
        [SchemaVerdict.CORRECT]: 0,
        [SchemaVerdict.INCORRECT]: 0,
        [SchemaVerdict.UNMARKED]: 0,
      };
    } else {
      this.sectionLevelStreak = 0;
      this.previousStreakVerdict = null;
    }
  }

  /**
   * Parse verdict marking from scheme configuration
   */
  private parseVerdictMarkingFromScheme(
    sectionScheme: SectionSchemeConfig | any
  ): Record<string, number> {
    const parsedMarking: Record<string, number> = {};

    for (const verdict of Object.values(Verdict)) {
      const schemaVerdict = AnswerMatcher.getSchemaVerdict(AnswerType.STANDARD, verdict, 0);

      // Try to get from the scheme
      let schemaVerdictMarking: number;
      if (sectionScheme[schemaVerdict] !== undefined) {
        schemaVerdictMarking = this.parseVerdictMarking(sectionScheme[schemaVerdict]);
      } else {
        // Default values
        schemaVerdictMarking = schemaVerdict === SchemaVerdict.CORRECT ? 1 : 0;
      }

      // Warn if positive marks for incorrect in non-bonus sections
      if (
        this.markingType === MarkingSchemeType.DEFAULT &&
        schemaVerdictMarking > 0 &&
        schemaVerdict === SchemaVerdict.INCORRECT &&
        !this.sectionKey.startsWith(BONUS_SECTION_PREFIX)
      ) {
        logger.warn(
          `Found positive marks (${schemaVerdictMarking.toFixed(2)}) for incorrect answer ` +
          `in schema '${this.sectionKey}'. For Bonus sections, prefer adding prefix 'BONUS_' to the scheme name.`
        );
      }

      parsedMarking[verdict] = schemaVerdictMarking;
    }

    return parsedMarking;
  }

  /**
   * Parse verdict marking (handles fractions and arrays for streaks)
   */
  private parseVerdictMarking(marking: any): number {
    if (Array.isArray(marking)) {
      // For streak-based marking, use first value as default
      return this.parseFloatOrFraction(marking[0]);
    }
    return this.parseFloatOrFraction(marking);
  }

  /**
   * Parse float or fraction string to number
   */
  private parseFloatOrFraction(value: string | number): number {
    if (typeof value === 'number') {
      return value;
    }
    if (value.includes('/')) {
      const [num, denom] = value.split('/').map(Number);
      return num / denom;
    }
    return parseFloat(value);
  }

  /**
   * Get delta for a specific verdict and streak
   */
  private getDeltaForVerdict(
    answerMatcherMarking: Record<string, number>,
    questionVerdict: string,
    currentStreak: number
  ): number {
    const marking = answerMatcherMarking[questionVerdict];

    if (Array.isArray(marking)) {
      // Return score based on streak index
      return marking[currentStreak] || marking[marking.length - 1];
    }

    if (currentStreak > 0) {
      logger.warn(
        `Non-zero streak (${currentStreak}) for verdict ${questionVerdict} in scheme ${this}. ` +
        `Using non-streak score for this verdict.`
      );
    }

    return marking;
  }

  /**
   * Get delta and update streak based on verdict
   */
  getDeltaAndUpdateStreak(
    answerMatcherMarking: Record<string, number>,
    answerType: AnswerType,
    questionVerdict: string,
    allowStreak: boolean
  ): {
    delta: number;
    currentStreak: number;
    updatedStreak: number;
  } {
    const schemaVerdict = AnswerMatcher.getSchemaVerdict(answerType, questionVerdict, 0);

    let currentStreak: number;
    let updatedStreak: number;
    let delta: number;

    if (this.markingType === MarkingSchemeType.VERDICT_LEVEL_STREAK) {
      currentStreak = this.streaks[schemaVerdict];

      // Reset all streaks
      this.resetAllStreaks();

      // Increase only current verdict streak
      if (allowStreak && schemaVerdict !== SchemaVerdict.UNMARKED) {
        this.streaks[schemaVerdict] = currentStreak + 1;
      }

      delta = this.getDeltaForVerdict(answerMatcherMarking, questionVerdict, currentStreak);
      updatedStreak = this.streaks[schemaVerdict];

    } else if (this.markingType === MarkingSchemeType.SECTION_LEVEL_STREAK) {
      currentStreak = this.sectionLevelStreak;
      const previousVerdict = this.previousStreakVerdict;

      // Reset
      this.resetAllStreaks();

      // Increase only if same verdict continues
      if (
        allowStreak &&
        (previousVerdict === null || schemaVerdict === previousVerdict)
      ) {
        this.sectionLevelStreak = currentStreak + 1;
      }

      this.previousStreakVerdict = schemaVerdict;
      delta = this.getDeltaForVerdict(answerMatcherMarking, questionVerdict, currentStreak);
      updatedStreak = this.sectionLevelStreak;

    } else {
      // Default - no streaks
      currentStreak = 0;
      updatedStreak = 0;
      delta = this.getDeltaForVerdict(answerMatcherMarking, questionVerdict, currentStreak);
    }

    return { delta, currentStreak, updatedStreak };
  }

  /**
   * Create a deep copy with different questions
   * Port of Python's deepcopy_with_questions method
   */
  deepcopyWithQuestions(questions: string[]): SectionMarkingScheme {
    const clone = deepClone(this);
    clone.updateQuestions(questions);
    return clone;
  }

  /**
   * Update questions for this scheme
   */
  updateQuestions(questions: string[]): void {
    this.questions = questions;
    this.validateMarkingScheme();
  }

  /**
   * Validate marking scheme configuration
   */
  private validateMarkingScheme(): void {
    // Validate streak-based marking schemes
    if (
      this.markingType === MarkingSchemeType.VERDICT_LEVEL_STREAK ||
      this.markingType === MarkingSchemeType.SECTION_LEVEL_STREAK
    ) {
      // Check if any marking values are arrays (for streak bonuses)
      const streakMarkings: Array<{ verdict: string; length: number }> = [];

      for (const [verdict, marking] of Object.entries(this.marking)) {
        if (Array.isArray(marking)) {
          streakMarkings.push({ verdict, length: marking.length });
        }
      }

      // If we have questions and streak markings, validate lengths
      if (this.questions && this.questions.length > 0 && streakMarkings.length > 0) {
        const maxPossibleStreak = this.questions.length;

        for (const { verdict, length } of streakMarkings) {
          if (length < maxPossibleStreak) {
            logger.warn(
              `Marking scheme '${this.sectionKey}': Verdict '${verdict}' has ${length} streak levels ` +
              `but section has ${maxPossibleStreak} questions. Consider adding more streak levels ` +
              `or the last level will be used for all remaining streaks.`
            );
          }
        }
      }
    }
  }

  /**
   * Get bonus type for this section
   */
  getBonusType(): string | null {
    if (
      this.markingType === MarkingSchemeType.VERDICT_LEVEL_STREAK ||
      this.marking[Verdict.NO_ANSWER_MATCH] <= 0
    ) {
      return null;
    }
    if (this.marking[Verdict.UNMARKED] > 0) {
      return 'BONUS_FOR_ALL';
    }
    if (this.marking[Verdict.UNMARKED] === 0) {
      return 'BONUS_ON_ATTEMPT';
    }
    return null;
  }

  toString(): string {
    return this.sectionKey;
  }
}

